4ian/GDevelop
# Add support for uploading resources for cloud project

This allows to upload resources:
- Resources are uploaded to the backend which gives an URL where to upload them. We hash the content to ensure integrity and other nice features in the backend.
- The option to use a resource from an URL was hidden by default behind an "advanced" toggle.
- Reworks the "save as" of the "StorageProviders" to call "onMoveResources", which is plugged by MainFrame to `ResourcesMover`. This helper does the job of looking at the old "StorageProvider", the new "StorageProvider", and call a function that does the move of the resources between these storage providers.
  - This means that this `ResourcesMover` has the knowledge of every storage provider (which is fine!). It works a bit as a matrix between storage providers. 
  - We have a similar `ResourcesFetcher` which allows storage providers to fetch some resources, like resolving urls (when loading a project from a URL) or download files locally (when saving/opening a project from a URL, like an example, into a local file, on the desktop app)
  - ~~This means it could also replace the old "ResourcesFetcher" which was used "only" to import URLs into local files (i.e: this is an operation that should be handled by `ResourcesMover` using the 
 `UrlStorageProvider`=>`LocalFileStorageProvider` function).~~
  - ~~It's very similar to ResourcesFetcher in the sense that it allows to report errors and retry if something bad happened. A good reason to try to remove ResourcesFetcher and use this ResourcesMover only.~~
- There were a few changes to allow PIXI and resources to be loaded with "credentials" (i.e: cookies) being sent. Both for the preview, the images shown in the editor and the export (and the save as).

At the end, for the end user, this looks like this:
- A choice to upload one or more ressources: <img width="521" alt="image" src="https://user-images.githubusercontent.com/1280130/186684037-78e68381-a55a-42d1-a69b-713a60fb0406.png">
- Resources already uploaded are de-duplicated.
- If you "Save As" to another cloud project, the resources are uploaded to the newly created project, so it just works.
- "Download a copy" will give you a zip file with the project json and resources, to be opened in the desktop app (and it should "just work").

Stories:
- [ResourcesMoverDialog](https://gdevelop-storybook.s3.amazonaws.com/feature/upload-cloud-project-resources/latest/index.html?path=/story/resourcemoverdialog--default)
- [FileToCloudProjectResourceUploader](https://gdevelop-storybook.s3.amazonaws.com/feature/upload-cloud-project-resources/latest/index.html?path=/story/resourceslist-filetocloudprojectresourceuploader--default)

Longer term (but maybe not that far):
- [In another PR] You can use Cloud projects in the desktop app .
  - This includes save from/to a local project.
- [In another PR] You can use the asset store for single files in the desktop app.

Still to do:
- [x] Partially replace ResourcesFetcher by ResourcesMover.
- [x] Rework of LocalCreation/BrowserCreation so that a new project made from an example is loaded by UrlStorageProvider, and then "saved as" immediately with your chosen storage provider. 
- [x] Remove/replace ResourcesFetcher.
- ~~Check to see if cloud projects would work "as is" on the desktop app.~~
  - No because of adaptations need for the export.
- ~~Check if it means we can add the "resource stores" to the ResourceSources on desktop.~~
  - Needs some adaptation to fetch the resources, some props to rework.
- [x] Verify if the PIXI upgrade is actually not needed.
- [x] Fix remaining TODOs
  - [x] Some tests?
  - [x] Organise assets and avoid conflicts in filenames in DownloadFile.
  - [x] Revert changes in newIDE/web-app/scripts/deploy-GDJS-Runtime.js and newIDE/app/src/GameEngineFinder/BrowserS3GDJSFinder.js
- [x] Update DownloadFileStorageProvider to download resources as local files (with well organised assets by types). Use the `useGenericRetryableProcessWithProgress`. 
- [x] Ensure the backend provides proper limitation to prevent abuses for user accounts/IPs.
- [x] Ensure Google Drive still works.

Manual test cases:
- Create blank project (browser and desktop app). Add a new resource. Save.
- Create project from example (browser and desktop app). [Save] Add a new resource. Save.
- Save as to a new project (browser and desktop app). Check resources are copied.
- Download File on the webapp, verify resources are there. Verify project still works.
